### PR TITLE
Fixes Issue 18116 - to!wchar([string, dstring]), and to!char([wstring, dstring]) don't compile

### PR DESCRIPTION
**Pull request by Dechcaudron**
Fix for issue [18116](https://issues.dlang.org/show_bug.cgi?id=18116). Mentioning @schveiguy for review, due to discussion [here](https://github.com/dlang/phobos/pull/5946)

### CODE REVIEW

File: std/conv.d [hunk 1]
```diff
@@ -835,8 +844,8 @@ if (!isImplicitlyConvertible!(S, T) &&
     class C : B, I, J {}
     class D : I {}
 
-    static foreach (m1; 0 .. 5) // enumerate modifiers
-    static foreach (m2; 0 .. 5) // ditto
+    foreach (m1; AliasSeq!(0,1,2,3,4)) // enumerate modifiers
+    foreach (m2; AliasSeq!(0,1,2,3,4)) // ditto
```
Comment (wilzbach) [inline std/conv.d]: Huh?

File: std/conv.d [hunk 2]
```diff
@@ -1324,7 +1333,7 @@ in
 {
     assert(radix >= 2 && radix <= 36);
 }
-do
+body
```
Comment (wilzbach) [inline std/conv.d]: Huh?

File: std/conv.d [hunk 3]
```diff
@@ -2454,7 +2495,7 @@ in
...
```
Comment (wilzbach) [inline std/conv.d]: Huh?

File: std/conv.d [hunk 4]
```diff
@@ -1619,16 +1628,6 @@ if (!isImplicitlyConvertible!(S, T) && isAssociativeArray!S &&
...
 
-@safe unittest
-{
-    import std.algorithm.comparison : equal;
-    import std.array : byPair;
-
-    int[int] a;
-    assert(a.to!(int[int]) == a);
-    assert(a.to!(const(int)[int]).byPair.equal(a.byPair));
-}
```
Comment (wilzbach) [inline std/conv.d]: Why did you delete this test?

File: std/conv.d [hunk 5]
```diff
@@ -1883,6 +1882,48 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+         to type T.
+
+Throws: ConvException if the input range contains more than
+        a single code point, or if the code point does not
+        fit into a code unit of type T.
+*/
+private T toImpl(T, S)(S value)
+if (isInputRange!S && isSomeChar!(ElementEncodingType!S) &&
+     isSomeChar!T && !is(typeof(parse!T(value))) &&
+     is(typeof(parse!double(value))))
```
Comment (wilzbach) [inline std/conv.d]: This looks weird. There shouldn't be a need to check for `parse!double`

File: std/conv.d [hunk 6]
```diff
@@ -1883,6 +1882,48 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+{
+    immutable dchar codepoint = parse!dchar(value);
+    if (!value.empty)
+        throw new ConvException(convFormat("Cannot convert \"%s\" to %s because it " ~
+                                           "contains more than a single code point.",
+                                           value, T.stringof));
```
Comment (wilzbach) [inline std/conv.d]: Nit: two empty lines -> remove one.

File: std/conv.d (hunk 2 above)
Comment (schveiguy) [reply std/conv.d]: It looks like potentially @Dechcaudron used a stale version of the repository to base his changes from?

Comment (wilzbach) [reply std/conv.d]: Hmm git is smart enough for this (it thinks in hunks/changes, not files). It looks more like a rebase with mistakes or copy/pasting an old version of this file into a up-to-date branch.

**Comment by schveiguy** [comment]
@Dechcaudron in addition to the changes already requested, you should change the log message in your commit to reference the issue fixed. i.e. say "Fixes issue 18116" somewhere in the log message. That triggers the automatic changelog editing.

File: std/conv.d (hunk 5 above)
Comment (Dechcaudron) [reply std/conv.d]: I use parse!double(value) within the method, and given the argument is of type S, it may not compile. I'll check and see if instantiation conditions for parse overlap the ones on S for this method,

File: std/conv.d (hunk 2 above)
Comment (Dechcaudron) [reply std/conv.d]: @wilzbach's assesment is accurate. I fucked up real bad with this one. I'll fix it and let you guys know.

File: std/conv.d (hunk 6 above)
Comment (Dechcaudron) [reply std/conv.d]: I don't see the empty lines. Do you mind pointing them out?

Comment (wilzbach) [reply std/conv.d]: Ah the GitHub mobile view didn't display the "contains more" lines. Sorry. Please ignore

**Comment by Dechcaudron** [comment]
Okay, everything up to code now (pun intended).

File: std/conv.d [hunk 7]
```diff
@@ -1883,6 +1883,47 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+private T toImpl(T, S)(S value)
+if (isSomeChar!T && !is(typeof(parse!T(value))) &&
+    is(typeof(parse!double(value))))
+{
+    immutable dchar codepoint = parse!dchar(value);
+    if (!value.empty)
+        throw new ConvException(convFormat("Cannot convert \"%s\" to %s because it " ~
+                                           "contains more than a single code point.",
+                                           value, T.stringof));
+    auto equivalentTArray = to!(T[])(codepoint);
```
Comment (schveiguy) [inline std/conv.d]: Use `std.utf.encode`, as `to!(T[])` will allocate.

```D
T[dchar.sizeof / T.sizeof] outbuf; // hacky sizing, but it's correct.
auto resultSize = encode(outbuf, codepoint);
if(resultSize != 1) { ... } // throw conv error.
return outbuf[0];
```

File: std/conv.d [hunk 8]
```diff
@@ -1883,6 +1883,47 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+    if (equivalentTArray.length > 1)
+        throw new ConvException(convFormat("Code point '%s' does not fit into a single " ~
+                                           "%s code unit", codepoint, T.stringof));
```
Comment (schveiguy) [inline std/conv.d]: Slightly adjust the message here, remember the input was a string, not a code point.

File: std/conv.d [hunk 9]
```diff
@@ -1883,6 +1883,47 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+    import std.exception : assertThrown;
+
+    assert(toImpl!wchar("a") == 'a');
+
+    assert(toImpl!char("a"d) == 'a');
+    assert(toImpl!char("a"w) == 'a');
+    assert(toImpl!wchar("a"d) == 'a');
+
+    assertThrown!ConvException(toImpl!wchar("ab"));
+    assertThrown!ConvException(toImpl!char("😃"d));
```
Comment (schveiguy) [inline std/conv.d]: looks good!

File: std/conv.d [hunk 10]
```diff
@@ -1883,6 +1883,47 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
...
+Returns: the first code point of the input range, converted
+         to type T.
+
+Throws: ConvException if the input range contains more than
+        a single code point, or if the code point does not
+        fit into a code unit of type T.
+*/
+private T toImpl(T, S)(S value)
+if (isSomeChar!T && !is(typeof(parse!T(value))) &&
+    is(typeof(parse!double(value))))
```
Comment (schveiguy) [inline std/conv.d]: What is this `parse!double` for? did you mean `parse!dchar`?

Comment (Dechcaudron) [reply std/conv.d]: Yep, my bad. Fixing.

File: std/conv.d (hunk 7 above)
Comment (Dechcaudron) [reply std/conv.d]: [As per the documentation](https://dlang.org/phobos/std_utf.html#encode) encode will throw UTFException if the input is not a valid UTF code point. Should I catch said exception and rethrow a ConvException?

File: std/conv.d (hunk 8 above)
Comment (Dechcaudron) [reply std/conv.d]: Updated, let's see if it's better now.

**Comment by wilzbach** [comment]
This isn't a regression. Please target master. The 2.078 release is about to happen soon and we should only merge regression fixes to it. You still need to rebase your PR locally as I just changed the target branch. If you need help, please let me/us know.

File: std/conv.d (hunk 7 above)
Comment (schveiguy) [reply std/conv.d]: Let's follow precedent -- whatever `to!wstring(string)` does with an invalid code point. I suspect it lets the exception through.

Comment (Dechcaudron) [reply std/conv.d]: Okay, no further changes then

Comment (schveiguy) [reply std/conv.d]: Yep, just tested, throws a `UTFException`.

**Comment by Dechcaudron** [comment]
My pleasure 😃. But why are the checks failing?

**Comment by schveiguy** [comment]
> But why are the checks failing?

See my commit, trailing whitespace 😄

**Comment by Dechcaudron** [comment]
Oh, so that's why. Thanks!

File: std/conv.d [hunk 11]
```diff
@@ -1883,6 +1883,49 @@ if (isInputRange!S && !isInfinite!S && isSomeChar!(ElementEncodingType!S) &&
     assert(to!int(byDchar("10"), 10) == 10);
 }
 
+/**
+String, or string-like input range, to char type not directly
```
Comment (JackStouffer) [inline std/conv.d]: Sorry to jump in after this is already been merged,  but this documentation will not be seen because this function is private. This should have gone as the docs to an example underneath to.

Comment (schveiguy) [reply std/conv.d]: Hm... the docs don't talk about `to!char(string)` which worked before this change (and was the impetus to finding this limitation). I suppose it wouldn't hurt to rearrange the unit test to be part of the public docs and add some more comments.

Comment (Dechcaudron) [reply std/conv.d]: The thing is, since all `to` does is call `toImpl`, all documentation of the latter does not show up in the generated docs. As a matter of fact, the unit tests I added do not call `to` directly, but rather `toImpl`, so there is not much sense in having them document `to`. But I do agree `to` could benefit from better documentation. Should that be a separate issue?

Comment (ZombineDev) [reply std/conv.d]: Sounds good

Comment (schveiguy) [reply std/conv.d]: @Dechcaudron would you mind refactoring your unit tests to call `to` instead of `toImpl` (I should have caught that), moving the unit test to be under the `to` function, making it documented and putting a little blurb about it? I think you could even add calls for `to!char(string)` to show that works also.

Comment (Dechcaudron) [reply std/conv.d]: I think it makes sense to duplicate the tests rather than to plain move them. After all, we still want this particular override of `toImpl` tested. If I'm going to do something that affects the public documentation of `to`, I want to put some more thought into it. I'll get back to you guys about this in some days. I need to disconnect from Phobos a little. 

Should I create a separate pull request once I've gone about the above? Shall we create a new issue then?

Comment (Dechcaudron) [reply std/conv.d]: Guys, ready to do this over here. New issue, or just a standalone PR?

Comment (JackStouffer) [reply std/conv.d]: just make a new PR, no need for a bugzilla issue
